### PR TITLE
feat(instrumentation/asgi): add target to metrics

### PR DESCRIPTION
**Pull request by sk-**
# Description
This PR adds the target information for metrics reported by instrumentation/asgi.

Unfortunately, there's no ASGI standard to reliably get this information, and I was only able to get it for FastAPI.

I also tried to get the info with Sanic and Starlette (https://github.com/encode/starlette/issues/685), but there's nothing in the scope allowing to recreate the route.

Note to reviewers: I tried to touch as less as possible, so that we don;t require a refactor before this change. However, we could consider changing `collect_request_attributes` so that it returns both a trace attributes and a metrics attributes. Wihout that change we cannot add the `HTTP_TARGET` attribute to the list of metric atttributes, because it will be present but with high cardinality.

Partially fixes #1116

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Besides the included unit tests, the logic was tested using the following app:

```python
import io
import fastapi

app = fastapi.FastAPI()

def dump_scope(scope):
    b = io.StringIO()
    print(scope, file=b)
    return b.getvalue()

@app.get("/test/{id}")
def test(id: str, req: fastapi.Request):
    print(req.scope)
    return {"target": _collect_target_attribute(req.scope), "scope": dump_scope(req.scope)}

sub_app = fastapi.FastAPI()

@sub_app.get("/test/{id}")
def sub_test(id: str, req: fastapi.Request):
    print(req.scope)
    return {"target": _collect_target_attribute(req.scope), "scope": dump_scope(req.scope)}

app.mount("/sub", sub_app)
```

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by sk-** [comment]
Not sure what to do regarding the failed lint checks

```
************* Module opentelemetry.instrumentation.asgi
instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:449:4: R0912: Too many branches (13/12) (too-many-branches)
************* Module tests.test_asgi_middleware
instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py:1:0: C0302: Too many lines in module (1036/1000) (too-many-lines)
```

* How can I test the feature if there are no lines left to be added?
* One branch could be removed from the method `__call__` by changing
```python
for key, value in attributes.items():
  current_span.set_attribute(key, value)
```
to
```python
current_span.set_attributes(attributes)
```

However, that's not even related to this PR.

I could also have a function that modifies the list of attributes, and hence moving the `if target` check to that function.

However, I think it'd be best to change `collect_request_attributes` to return a set of attributes suitable for metrics and another for traces. However, that would require a refactoring several places, as that function is public and actively used today.

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `opentelemetry-instrumentation-grpc` add supports to filter requests to instrument. ([#1241](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1241))
 - Flask sqlalchemy psycopg2 integration
   ([#1224](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1224))
+- `opentelemetry-instrumentation-asgi` metrics record target attribute (FastAPI only)
```
Comment (srikanthccv) [inline CHANGELOG.md:15]: move this to unrelated section

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 2]
```diff
@@ -365,6 +365,34 @@ def get_default_span_details(scope: dict) -> Tuple[str, dict]:
...
+
+    Note: this function requires specific code for each framework, as there's no
+    standard attribute to use.
+    """
+    # FastAPI
+    root_path = scope.get("root_path", "")
+
+    route = scope.get("route")
+    if not route:
+        return None
```
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:388]:
```suggestion
```
redundant?

File: instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py [hunk 3]
```diff
@@ -705,6 +736,36 @@ def test_credential_removal(self):
...
+        self.scope["root_path"] = "/api/v2"
+        self.assertEqual(
+            otel_asgi._collect_target_attribute(self.scope),
+            "/api/v2/users/{user_id}",
+        )
+
+    def test_collect_target_attribute_fastapi_starlette_invalid(self):
+        self.scope["route"] = object()
+        self.assertIsNone(
+            otel_asgi._collect_target_attribute(self.scope), None
```
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py:766]:
```suggestion
            otel_asgi._collect_target_attribute(self.scope), "HTTP_TARGET values is not None"
```

File: CHANGELOG.md (hunk 1 above)
Comment (sk-) [reply CHANGELOG.md:15]: What do you mean by this? I did not found any unrelated sections?

The entry is already in the `Unreleased` section.

Comment (srikanthccv) [reply CHANGELOG.md:15]: Ah sorry, that was a typo. 

> The entry is already in the Unreleased section.

Not if you resolve the conflicts. Please resolve conflicts and move to the unreleased section.

Comment (sk-) [reply CHANGELOG.md]: @srikanthccv updated, please take another look

**Review by srikanthccv — APPROVED** [review]

**Comment by sk-** [comment]
@srikanthccv any advice regarding what to do to fix the lint violations?

See https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1323#issuecomment-1242810964

It's complaining about too many lines in the test file and too many branches in the main function.

**Comment by srikanthccv** [comment]
Hm, I am not sure. I don't mind if you add pylint-disable.

**Comment by sk-** [comment]
> Hm, I am not sure. I don't mind if you add pylint-disable.

Done, let me know if everything looks good now.

**Pull request merged**